src/server/bitops_family.cc

```diff
@@ -0,0 +1,285 @@
+#include "server/error.h"
+#include "server/tiered_storage.h"
+#include "server/transaction.h"
+#include "util/varz.h"
+
+namespace dfly {
+using namespace facade;
+
+namespace {
+static const int32_t OFFSET_FACTOR = 8;  // number of bits in byte
+// The following is the list of the functions that would handle the
+// commands that handle the bit operations
+void BitPos(CmdArgList args, ConnectionContext* cntx);
+void BitCount(CmdArgList args, ConnectionContext* cntx);
+void BitField(CmdArgList args, ConnectionContext* cntx);
+void BitFieldRo(CmdArgList args, ConnectionContext* cntx);
+void BitOp(CmdArgList args, ConnectionContext* cntx);
+void GetBit(CmdArgList args, ConnectionContext* cntx);
+void SetBit(CmdArgList args, ConnectionContext* cntx);
+OpResult<std::string> ReadValue(const OpArgs& op_args, std::string_view key);
+std::string GetString(EngineShard* shard, const PrimeValue& pv);
+// ------------------------------------------------------------------------- //
+//  Bits manipulation functions
+constexpr int32_t GetBitIndex(int32_t offset) noexcept {
+  return offset % OFFSET_FACTOR;
+}
+
+constexpr int32_t GetNormalizedBitIndex(int32_t offset) noexcept {
+  return (OFFSET_FACTOR - 1) - GetBitIndex(offset);
+}
+
+constexpr int32_t GetByteIndex(int32_t offset) noexcept {
+  return offset / OFFSET_FACTOR;
+}
+
+uint8_t GetByteValue(const std::string& str, int32_t offset) {
+  return static_cast<uint8_t>(str[GetByteIndex(offset)]);
+}
+
+constexpr bool CheckBitStatus(uint8_t byte, int32_t offset) {
+  return byte & (0x1 << offset);
+}
+
+// return true if bit is on
+bool GetBitValue(const std::string& entry, int32_t offset) {
+  const auto byte_val{GetByteValue(entry, offset)};
+  const auto index{GetNormalizedBitIndex(offset)};
+  return CheckBitStatus(byte_val, index);
+}
+
+bool GetBitValueSafe(const std::string& entry, int32_t offset) {
+  return ((entry.size() * OFFSET_FACTOR) > (std::size_t)offset) ? GetBitValue(entry, offset)
+                                                                : false;
+}
+
+constexpr uint8_t TurnBitOn(uint8_t on, int32_t offset) {
+  return on |= 1 << offset;
+}
+
+constexpr uint8_t TunBitOff(uint8_t on, int32_t offset) {
+  return on &= ~(1 << offset);
+}
+
+bool SetBitValue(std::string& entry, int32_t offset, bool bit_value) {
+  // we need to return the old value after setting the value for offset
+  const auto old_value{GetBitValue(entry, offset)};  // save this as the return value
+  auto byte{GetByteValue(entry, offset)};
+  std::bitset<8> bits{byte};
+  const auto bit_index{GetNormalizedBitIndex(offset)};
+  byte = bit_value ? TurnBitOn(byte, bit_index) : TunBitOff(byte, bit_index);
+  entry[GetByteIndex(offset)] = byte;
+  return old_value;
+}
+
+// ------------------------------------------------------------------------- //
+// Helper functions to access the data or change it
+
+void UpdateBitMapValue(DbIndex index, DbSlice& db_slice, PrimeIterator& it, std::string_view key,
```

Comment:
also you use this function only within OverrideValue::Set . 
why not make it private -> then you won't need to pass `index` and `db_slice `

Reply:
db slice is not a member, and I think that down the road this may need to be used outside (I did not finish implementing the commands yet). So maybe we would wait with moving it? I don't really like this, but the best solution would be to make it a lambda, but again I think this maybe useful later for other cases as well